shell/platform/android/io/flutter/embedding/android/FlutterFragment.java

```diff
@@ -729,6 +732,259 @@ public <T extends FlutterFragment> T build() {
     }
   }
 
+  /**
+   * Returns a {@link NewEngineInGroupFragmentBuilder} to create a {@code FlutterFragment} with a
+   * cached {@link io.flutter.embedding.engine.FlutterEngineGroup} in {@link
+   * io.flutter.embedding.engine.FlutterEngineGroupCache}.
+   *
+   * <p>An {@code IllegalStateException} will be thrown during the lifecycle of the {@code
+   * FlutterFragment} if a cached {@link io.flutter.embedding.engine.FlutterEngineGroup} is
+   * requested but does not exist in the {@link
+   * io.flutter.embedding.engine.FlutterEngineGroupCache}.
+   */
+  @NonNull
+  public static NewEngineInGroupFragmentBuilder withNewEngineInGroup(
+      @NonNull String engineGroupId) {
+    return new NewEngineInGroupFragmentBuilder(engineGroupId);
+  }
+
+  /**
+   * Builder that creates a new {@code FlutterFragment} that uses a cached {@link
+   * io.flutter.embedding.engine.FlutterEngineGroup} to create a new {@link
+   * io.flutter.embedding.engine.FlutterEngine} with {@code arguments} that correspond to the values
+   * set on this {@code Builder}.
+   *
+   * <p>Subclasses of {@code FlutterFragment} that do not introduce any new arguments can use this
+   * {@code Builder} to construct instances of the subclass without subclassing this {@code
+   * Builder}. {@code MyFlutterFragment f = new
+   * FlutterFragment.NewEngineInGroupFragmentBuilder(MyFlutterFragment.class, engineGroupId)
+   * .someProperty(...) .someOtherProperty(...) .build<MyFlutterFragment>(); }
+   *
+   * <p>Subclasses of {@code FlutterFragment} that introduce new arguments should subclass this
+   * {@code NewEngineInGroupFragmentBuilder} to add the new properties:
+   *
+   * <ol>
+   *   <li>Ensure the {@code FlutterFragment} subclass has a no-arg constructor.
+   *   <li>Subclass this {@code NewEngineInGroupFragmentBuilder}.
+   *   <li>Override the new {@code NewEngineInGroupFragmentBuilder}'s no-arg constructor and invoke
+   *       the super constructor to set the {@code FlutterFragment} subclass: {@code public
+   *       MyBuilder() { super(MyFlutterFragment.class); } }
+   *   <li>Add appropriate property methods for the new properties.
+   *   <li>Override {@link NewEngineInGroupFragmentBuilder#createArgs()}, call through to the super
+   *       method, then add the new properties as arguments in the {@link Bundle}.
+   * </ol>
+   *
+   * Once a {@code NewEngineInGroupFragmentBuilder} subclass is defined, the {@code FlutterFragment}
+   * subclass can be instantiated as follows. {@code MyFlutterFragment f = new MyBuilder()
+   * .someExistingProperty(...) .someNewProperty(...) .build<MyFlutterFragment>(); }
+   */
+  public static class NewEngineInGroupFragmentBuilder {
```

Comment:
There is so much duplicated here vis-a-vis the other builders.  Is there a way we can share some of this?

Reply:
I also considering this problem, now we have too many builder, `NewEngineIntentBuilder`, `CachedEngineIntentBuilder`, as well as the increase of `NewEngineInGroupIntentBuilder`. They have some common parameters (especially NewEngineIntentBuilder and NewEngineInGroupIntentBuilder, because they are starting a new engine), in FlutterActivity, FlutterFragmentActivity, FlutterFragment all appear.

Reply:
I had a rough idea: could I abstract a Builder class with some parameters used by Dart like entryPoint, initialRoute, dartEntrypointArgs. Then Actvitity or fragments of the Builder extends from this, and contains specific attributes, such as activityClass/fragmentClass. Many of such NewEngineIntentBuilder and NewEngineInGroupIntentBuilder fields can be Shared. What do you think? I would probably do it in a new PR if I could.

Reply:
Doing it in a new PR sounds fine, please file an issue.

Reply:
OK  https://github.com/flutter/flutter/issues/114567